packages/connection-router-api/src/primitives.rs

```diff
@@ -0,0 +1,187 @@
+    /// 1. it must be possible to use its content to find the corresponding message on the source chain
+    /// 2. the ID must uniquely identify the message on the source chain, i.e. no two messages can have the same ID, and no single message can have multiple valid IDs.
+    ///
+    /// IMPORTANT: Verifier contracts must enforce these requirements.
+    pub id: String,
+    pub source_chain: ChainName,
+    pub source_address: Address,
+    pub destination_chain: ChainName,
+    pub destination_address: Address,
+    /// hash length is enforced to be 32 bytes
+    pub payload_hash: [u8; 32],
+}
+
+// [cw_serde] has been expanded here because we need to implement PartialEq manually
+#[derive(
+    ::cosmwasm_schema::serde::Serialize,
+    ::cosmwasm_schema::serde::Deserialize,
+    Clone,
+    Debug,
+    ::cosmwasm_schema::schemars::JsonSchema,
+)]
+#[serde(deny_unknown_fields, crate = "::cosmwasm_schema::serde")]
+#[schemars(crate = "::cosmwasm_schema::schemars")]
+#[serde(try_from = "String")]
+pub struct ChainName(String);
+
+impl Hash for ChainName {
+    /// this is implemented manually because we want to ignore case when hashing
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.0.to_lowercase().hash(state)
+    }
+}
+
+impl PartialEq for ChainName {
+    /// this is implemented manually because we want to ignore case when checking equality
+    fn eq(&self, other: &Self) -> bool {
+        self.0.to_lowercase() == other.0.to_lowercase()
+    }
+}
+
+impl FromStr for ChainName {
+    type Err = Error;
+
+    fn from_str(chain_name: &str) -> Result<Self, Self::Err> {
+        fn matches_chain_name_pattern(s: &str) -> bool {
+            Regex::new(CHAIN_NAME_REGEX)
+                .expect("invalid regex pattern for chain name")
+                .is_match(s)
+        }
+
+        if matches_chain_name_pattern(chain_name) {
+            Ok(ChainName(chain_name.to_string()))
+        } else {
+            Err(Error::ChainNamePatternMismatch(chain_name.to_string()))
+        }
+    }
+}
+
+impl TryFrom<String> for ChainName {
+    type Error = Error;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        value.parse()
+    }
+}
+
+impl Display for ChainName {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        Display::fmt(&self.0, f)
+    }
+}
+
+#[cw_serde]
+pub struct Address(nonempty::String);
+
+impl FromStr for Address {
+    type Err = Report<Error>;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        s.parse::<nonempty::String>()
+            .change_context(Error::EmptyAddress)
+            .map(Address)
+    }
+}
+impl TryFrom<String> for Address {
+    type Error = Report<Error>;
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        value.parse()
+    }
+}
+
+impl Deref for Address {
+    type Target = str;
+
+    fn deref(&self) -> &Self::Target {
+        self.0.deref()
```

Comment:
```suggestion
        self.0
```

Reply:
doesn't work, Addr is a nonempty::String, so I need to deref that